msal/src/main/java/com/microsoft/identity/client/SingleAccountPublicClientApplication.java

```diff
@@ -95,4 +100,114 @@ public void onError(Exception exception) {
+                appContext,
+                accountCredentialCache,
+                new MicrosoftStsAccountCredentialAdapter());
+    }
+
+    private IAccountCredentialCache getCurrentAccountCredentialCache(final Context appContext){
+        return new SharedPreferencesAccountCredentialCache(
+                new CacheKeyValueDelegate(),
+                new SharedPreferencesFileManager(
+                        appContext,
+                        SINGLE_ACCOUNT_CREDENTIAL_SHARED_PREFERENCES,
+                        new StorageHelper(appContext)
+                )
+        );
+    }
+
+    /**
+     * Get current account that is persisted in shared preference.
+     *
+     * @param appContext
+     * @return a persisted MultiTenantAccount. This could be null.
+     * */
+    @Nullable
+    private MultiTenantAccount getPersistedCurrentAccount(@NonNull final Context appContext) {
+        final String methodName = ":getPersistedCurrentAccount";
+        final IAccountCredentialCache accountCredentialCache = getCurrentAccountCredentialCache(appContext);
+        final MsalOAuth2TokenCache msalOAuth2TokenCache = getTokenCache(appContext, accountCredentialCache);
+
+        List<AccountRecord> accountRecords = accountCredentialCache.getAccounts();
+        if (accountRecords.size() == 0) {
+            return null;
+        } else if (accountRecords.size() > 1) {
+            com.microsoft.identity.common.internal.logging.Logger.verbose(
+                    TAG + methodName,
+                    "Returned accountRecords are larger than 1. " +
+                            "This is unexpected in Single account mode." +
+                            "Clear the cache and return null.");
+
+            accountCredentialCache.clearAll();
+            return null;
+        }
+
+        AccountRecord accountRecord = accountRecords.get(0);
+        List<IdTokenRecord> idTokenRecordList = msalOAuth2TokenCache.getIdTokensForAccountRecord(null, accountRecord);
+
+        List<ICacheRecord> cacheRecordList = new ArrayList<>();
+        for (final IdTokenRecord idTokenRecord : idTokenRecordList) {
+            CacheRecord cacheRecord = new CacheRecord();
+            cacheRecord.setAccount(accountRecord);
+            cacheRecord.setIdToken(idTokenRecord);
+            cacheRecordList.add(cacheRecord);
+        }
+
+        return getAccountFromICacheRecordList(cacheRecordList);
+    }
+
+    /**
+     * Persists current account to shared preference.
+     *
+     * @param appContext
+     * @param cacheRecords list of cache record that belongs to an account.
+     *                     Please note that this layer will not verify if the list belongs to a single account or not.
+     * */
+    private void persistCurrentAccount(@NonNull final Context appContext,
+                                       @Nullable List<ICacheRecord> cacheRecords) {
+        final IAccountCredentialCache accountCredentialCache = getCurrentAccountCredentialCache(appContext);
```

Comment:
Same here, `IAccountCredentialCache` should not be used directly -- if there are APIs you need here that aren't available on `MsalOauth2TokenCache` let me know and we can add additional APIs